dapr/proto/runtime/v1/dapr.proto

```diff
@@ -121,6 +121,18 @@ message InvokeBindingRequest {
   // have a default time to live. The message ttl overrides any value
   // in the binding definition.
   map<string,string> metadata = 3;
+
+  // The name of the operation type for the binding to invoke
+  string operation = 4;

```

Comment:
Can we define enums for operation?

Reply:
No, there's no defined list for this. It's dynamic.

Reply:
Even if it is dynamic, we already set the number of operations in component-contrib. User needs to know which operations are supported in component contrib.

So, isn't it better to define the list of common operations that can be shared across componen-contrib? It will also give both end-user and component-contrib contributers the clear message for the supported operations. We will not be unable to remove or change the operation of the list, but we can add more operation dynamically as we need. 

cc/ @amanbha @msfussell @orizohar

Reply:
@youngbupark that's not correct.. there are basic operations in components-contrib to help contributors use the same ones when needed, but the list is actually dynamic, and the value can be anything.

New components can come up with any number of operations, and the list in components-contrib will only grow if needed.

Reply:
Using enums here will greatly hurt maintainability, as it will become a total sum of every new operation that gets added to any component.

In addition, it will introduce a tight coupling between two things that should never be coupled: the Dapr API for the runtime, and the component specific implementation in `contrib`.

Reply:
In many cases, for maintainability of enum, we can describe the process to add new operation as a comment.

As a user perspective, more strict typing can give the clear message/explanation instead of plain-text as the project grows. so I prefer defining enums or any other stricter operation type.

Reply:
Like I said, the maintainability will suffer greatly. Giving a process will not change the fact that the API will be tightly coupled to all of the implementations in contrib, and any change to components contrib will require the following:

1. Change code in contrib
2. Change the Dapr proto API
3. Regenerate the gRPC files *for every language we have an SDK for*
4. Update the SDKs with the new proto files

We have the API documentation for bindings to make it explicit and clear which operations are supported for which binding.
Even if the enums are hard-coded inside the proto service definition, the user still needs to look at the docs and see which operation is supported for a given binding, because not all bindings support all operations.

Reply:
Leaving it as string keeps it easily extensible and no changes in protos and sdks needed.
Http calls will anyhow take it as a string.
If bindings can return correct error message after checking for operations, it would be a good balance.

Reply:
Yep, the runtime validates if the operation is supported or not.

https://github.com/dapr/dapr/pull/1608/files#diff-a6a78e03551b53fec882ef23c9762278R505

Reply:
> Like I said, the maintainability will suffer greatly. Giving a process will not change the fact that the API will be tightly coupled to all of the implementations in contrib, and any change to components contrib will require the following:
> 
> 1. Change code in contrib
> 2. Change the Dapr proto API
> 3. Regenerate the gRPC files _for every language we have an SDK for_
> 4. Update the SDKs with the new proto files
> 
> We have the API documentation for bindings to make it explicit and clear which operations are supported for which binding.
> Even if the enums are hard-coded inside the proto service definition, the user still needs to look at the docs and see which operation is supported for a given binding, because not all bindings support all operations.

I missed updating SDK step :) That make sense.